gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java

```diff
@@ -0,0 +1,275 @@
+      throw new RuntimeException(roe);
+    } catch (IOException ioe) {
+      log.error(String.format("Failed to commit dataset state for dataset %s of job %s", this.datasetUrn,
+          this.jobContext.getJobId()), ioe);
+      throw new RuntimeException(ioe);
+    } finally {
+      try {
+        finalizeDatasetState(datasetState, datasetUrn);
+
+        if (commitSequenceBuilder.isPresent()) {
+          buildAndExecuteCommitSequence(commitSequenceBuilder.get(), datasetState, datasetUrn);
+          datasetState.setState(JobState.RunningState.COMMITTED);
+        } else {
+          persistDatasetState(datasetUrn, datasetState);
+        }
+      } catch (IOException | RuntimeException ioe) {
+        log.error(
+            String.format("Failed to persist dataset state for dataset %s of job %s", datasetUrn, this.jobContext.getJobId()), ioe);
+        throw new RuntimeException(ioe);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Synchronized version of {@link #commitDataset(JobState.DatasetState, DataPublisher)} used when publisher is not
+   * thread safe.
+   */
+  private synchronized void safeCommitDataset(JobState.DatasetState datasetState, DataPublisher publisher) {
+    commitDataset(datasetState, publisher);
+  }
+
+  /**
+   * Commit the output data of a dataset.
+   */
+  private void commitDataset(JobState.DatasetState datasetState, DataPublisher publisher) {
+
+    try {
+      publisher.publish(datasetState.getTaskStates());
+    } catch (Throwable t) {
+      log.error("Failed to commit dataset", t);
+      setTaskFailureException(datasetState.getTaskStates(), t);
+    }
+
+    // Set the dataset state to COMMITTED upon successful commit
+    datasetState.setState(JobState.RunningState.COMMITTED);
+  }
+
+  private synchronized void buildAndExecuteCommitSequence(CommitSequence.Builder builder, JobState.DatasetState datasetState,
+      String datasetUrn) throws IOException {
+    CommitSequence commitSequence =
+        builder.addStep(buildDatasetStateCommitStep(datasetUrn, datasetState).get()).build();
+    this.jobContext.getCommitSequenceStore().get().put(commitSequence.getJobName(), datasetUrn, commitSequence);
+    commitSequence.execute();
+    this.jobContext.getCommitSequenceStore().get().delete(commitSequence.getJobName(), datasetUrn);
+  }
+
+  /**
+   * Finalize a given {@link JobState.DatasetState} before committing the dataset.
+   *
+   * This method is thread-safe.
+   */
+  private void finalizeDatasetStateBeforeCommit(JobState.DatasetState datasetState) {
+    for (TaskState taskState : datasetState.getTaskStates()) {
+      if (taskState.getWorkingState() != WorkUnitState.WorkingState.SUCCESSFUL
+          && this.jobContext.getJobCommitPolicy() == JobCommitPolicy.COMMIT_ON_FULL_SUCCESS) {
+        // The dataset state is set to FAILED if any task failed and COMMIT_ON_FULL_SUCCESS is used
+        datasetState.setState(JobState.RunningState.FAILED);
```

Comment:
Note all of this code is just moved from `JobContext`. I can fix it if necessary, but in principle this is outside of the scope of this PR.